### ISSUE TITLE
Data loss while writing values with rate greater than heartbeat

### ISSUE THREAD

**Issue by pkunilov** [oetiker/rrdtool-1.x#395]
Hello, 
I do the following steps:

1. Create rrd db with sampling rate 1 sec and one datasource with type GAUGE and heartbeat 2 sec
2. Start writing values with rate 3 sec
3. Only one first value will be written all other ones will be lost and replaced with NaN.

For example, I write 10 values with rate 3 sec (in rrd db with settings described above)

Current output:
time1  Value1
time2  NaN
time3  NaN
time4  NaN
time5  NaN
time6  NaN
time7  NaN
time8  NaN
time9  NaN
time10 NaN

Expected Output:
time1  Value1
time2  NaN
time3  NaN
time4  Value2
time5  NaN
time6  NaN
time7  Value3
time8  NaN
time9  NaN
time10 Value4

I use RRDtool 1.2.15
The issue is reproduced on 
Windows 7
Linux 2.6.32-358.6.1.el6.x86_64

For example, rrdtool commands which can reproduce the above behavior:

rrdtool create "./test.rrd"  --start 1369748048 --step 1 DS:test:GAUGE:2:U:U RRA:MIN:0.999:1:30
rrdtool update "./test.rrd" 1369748049:1.0
rrdtool update "./test.rrd" 1369748052:2.0
rrdtool update "./test.rrd" 1369748055:3.0
rrdtool update "./test.rrd" 1369748058:4.0
rrdtool update "./test.rrd" 1369748061:5.0
rrdtool update "./test.rrd" 1369748064:6.0
rrdtool update "./test.rrd" 1369748067:7.0
rrdtool update "./test.rrd" 1369748070:8.0
rrdtool update "./test.rrd" 1369748073:9.0
rrdtool update "./test.rrd" 1369748076:10.0
rrdtool dump ./test.rrd

Is it a correct behavior? If it is correct, if there is any way to avoid data loss?

Best regards,
Peter Kunilov

**Comment by themuvarov** [comment]
The issue is reproduced on the latest released version v1.5.0-rc2.

tried 1.3.8, 1.4.7 and eventually https://github.com/oetiker/rrdtool-1.x/tree/v1.5.0-rc2

Enviroment: 2.6.32-504.el6.x86_64

Steps to reproduce:
rrdtool create ./test1_5_0.rrd --start 1369748048 --step 1 DS:test:GAUGE:2:U:U RRA:MIN:0.999:1:30

rrdtool update ./test1_5_0.rrd 1369748049:1.0
rrdtool update ./test1_5_0.rrd 1369748052:2.0
rrdtool update ./test1_5_0.rrd 1369748055:3.0
rrdtool update ./test1_5_0.rrd 1369748058:4.0
rrdtool dump ./test1_5_0.rrd

Results:

<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE rrd SYSTEM "http://oss.oetiker.ch/rrdtool/rrdtool.dtd">

<rrd>
	<version>0003</version>
	<step>1</step> 
	<lastupdate>1369748058</lastupdate> 

	<ds>
		<name> test </name>
		<type> GAUGE </type>
		<minimal_heartbeat>2</minimal_heartbeat>
		<min>NaN</min>
		<max>NaN</max>

		
		<last_ds>4.0</last_ds>
		<value>NaN</value>
		<unknown_sec> 0 </unknown_sec>
	</ds>

	
	<rra>
		<cf>MIN</cf>
		<pdp_per_row>1</pdp_per_row> 

		<params>
		<xff>9.9900000000e-01</xff>
		</params>
		<cdp_prep>
			<ds>
			<primary_value>NaN</primary_value>
			<secondary_value>NaN</secondary_value>
			<value>NaN</value>
			<unknown_datapoints>0</unknown_datapoints>
			</ds>
		</cdp_prep>
		<database>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>1.0000000000e+00</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
			 <row><v>NaN</v></row>
		</database>
	</rra>
</rrd>

Only the first measurement 1369748049 has been saved it DB. Others were lost.

**Comment by oetiker** [comment]
Well, if you have a required heartbeat of 2 seconds, but you update every 3 seconds, you will get unknown data (that is what the heartbeat is for, to catch long intervals).

The reason your first update works, is that you create at 1369748048 and update at 1369748049 which is only 1 second apart and with a GAUGE type datasource, one update is all you need for a valid entry. (with COUNTER or DERIVE you would need two obviously).

**Comment by themuvarov** [comment]
Please confirm that loosing first result after long interval is a design intent? 

It there a way to avoid loss?

**Comment by oetiker** [comment]
Yes this is BY DESIGN. But I agree, that if you use DS Type GAUGE one could argue, that you can backfill data up to mrhb into the past from the current point in time. Would you like to work on such a feature?

**Comment by oetiker** [comment]
### Re-triage: fixable for GAUGE

Reviewing this with current eyes: the original ``invalid`` closure was over-conservative. The behaviour is correct for the rate-typed DSes (COUNTER/DERIVE/ABSOLUTE/DCOUNTER/DDERIVE), but unnecessarily punitive for GAUGE.

In ``src/rrd_update.c`` the heartbeat gate is a single shared check (around line 1458):

```c
if ((updvals[ds_idx + 1][0] != 'U') &&
    (dst_idx != DST_CDEF) &&
    rrd->ds_def[ds_idx].par[DS_mrhb_cnt].u_cnt >= interval) {
```

When ``interval > heartbeat``, the GAUGE branch is skipped along with the rate-typed DSes. But the semantics aren't symmetric:

- **Rate-typed DSes:** the value is a rate over the interval. If too much time has passed since the last sample, there is no defensible rate — NaN'ing is correct.
- **GAUGE:** the value is a point measurement at the current timestamp. The heartbeat tells us the gap to the *previous* sample is suspicious; it doesn't make the current measurement meaningless.

A reasonable fix for GAUGE when ``interval > heartbeat``:

1. Mark the PDPs between ``last_update_time`` and ``(now − heartbeat)`` as unknown (the gap is genuinely unknown).
2. Still record ``newval`` into the PDPs covering the last ``heartbeat`` seconds, using a shortened spread of length ``heartbeat`` rather than ``interval``.

That delivers the reporter's expected behaviour — sparse-but-not-empty data — without lying about the gap.

Caveats:
- Behaviour change for GAUGE; users relying on the current ""whole-gap-is-NaN"" output would see different graphs. Worth a NEWS entry.
- Only relevant when ``step ≤ heartbeat < update_interval``.
- Does not extend to DCOUNTER/DDERIVE (those are rates).

Leaving open with ``bug`` + ``enhancement``.
